src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java

```diff
@@ -147,12 +147,12 @@ public void ready(boolean accept, boolean connect, boolean read, boolean write)
                             while (recv.hasRemaining()) {
                                 onRead(recv);
                             }
-                            if (in.isOpen() && recvKey.isValid()) {
-                                getIoHub().addInterestRead(recvKey);
-                            } else {
-                                recvKey.cancel();
-                                onRecvClosed();
-                            }
+                        }
+                        if (in.isOpen() && recvKey.isValid()) {
```

Comment:
```
if (count != -1) {
  ...
}
```

and hoist the scope of count outside the while loop... otherwise a count of -1 will trigger two closed events